### ISSUE TITLE
mut_from_ref should not trigger on code which is fully safe

### ISSUE THREAD

**Issue by jasoncarr0** [rust-lang/rust-clippy#6326]
The reason that is given for the lint is that "This is trivially unsound, as one can create two mutable references from the same (immutable!) source. This error actually lead to an interim Rust release 1.15.1." but this warning only makes sense when some of the code is unsafe.

This can occur without issue when using arenas, at least some of which follow the condition of producing an exclusive reference to new data using the shared reference to the arena.

```
$ cargo clippy -V                                                                                                                                      ~/code/rust/lang
clippy 0.0.212 (5404efc 2020-11-11)
```
```
$ rustc -Vv
rustc 1.49.0-nightly (5404efc28 2020-11-11)
binary: rustc
commit-hash: 5404efc28a0cddee103ef6396c48ea71ff9631c8
commit-date: 2020-11-11
host: x86_64-unknown-linux-gnu
release: 1.49.0-nightly
```

**Comment by llogiq** [comment]
In that case, the arena implements mutual exclusion without a Mutex or RwLock?

**Comment by jasoncarr0** [comment]
@llogiq it does not matter how the arena implements this. For me, this error occurred as a _user_ of an Arena type.

Both bumpalo and typed-arena use !Sync primitives (Cell/RefCell), rather than Mutex/RwLock.

**Comment by jasoncarr0** [comment]
Ah, okay I think I understand your question. Mutual exclusion is guaranteed because the data is disjoint and the unique reference to it is given at allocation.

**Comment by llogiq** [comment]
Why would a *user* of an arena get this error? Can you show or link a code example?

**Comment by jasoncarr0** [comment]
I think the unintuitiveness you're hitting on is exactly why I was surprised to see this. :grinning: 
It triggers because this triggers only based on interface: if you return a &mut reference with same lifetime as a & shared source and no mutable reference is used to create it, then this will trigger, regardless of what means you got the reference from.

I have some code which uses an arena to allocate some DAGs. For simplicity we'll just make it into a linked list of `usize`s, and I'll use typed arena.

```
type Arena<'alloc> = typed_arena::Arena<List<'alloc>>;
struct List<'alloc> {
  val: usize,
  rest: Option<&'a alloc>
}
impl<'alloc> List<'alloc> {
  pub fn new() -> Self { List { val: 0usize, rest: None }}
  pub fn push<'a>(&'alloc self, arena: &'a Arena<'alloc>, val: usize) -> &'a mut Self {
    alloc.alloc(List {
      val, rest: Some(self)
    })
  }
}
```
Note that it does not matter whether it takes &self, or &mut self, as I originally had in my code. In both cases, this lint triggers because of returning `&'a mut Anything` where `'a` is only used for a shared reference to the arena.

**Comment by llogiq** [comment]
Where does `alloc.alloc` in your example come from? Or would it need to be `arena.alloc`? I have a hunch we could fix this false positive by checking within the function if there is a *method call* that does the conversion.

**Comment by jasoncarr0** [comment]
Sorry, typo. it's arena.alloc. I renamed to be a bit more clear and missed that
